dbos/_sys_db.py

```diff
@@ -1112,88 +1113,153 @@ def start_queued_workflows(self, queue: "Queue", executor_id: str) -> List[str]:
+                dequeued_ids: List[str] = [row[0] for row in rows if row[1] is None][
+                    :max_tasks_this_worker_can_dequeue
+                ]
+                dbos_logger.debug(
+                    f"[{queue.name}] dequeueing {len(dequeued_ids)} task(s)"
                 )
-                ret_ids.append(id)
+                for id in dequeued_ids:
+
+                    # If we have a limiter, stop starting functions when the number
+                    # of functions started this period exceeds the limit.
+                    if queue.limiter is not None:
+                        if len(ret_ids) + num_recent_queries >= queue.limiter["limit"]:
+                            break
+
+                    # To start a function, first set its status to PENDING and update its executor ID
+                    c.execute(
+                        SystemSchema.workflow_status.update()
+                        .where(SystemSchema.workflow_status.c.workflow_uuid == id)
+                        .where(
+                            SystemSchema.workflow_status.c.status
+                            == WorkflowStatusString.ENQUEUED.value
+                        )
+                        .values(
+                            status=WorkflowStatusString.PENDING.value,
+                            executor_id=executor_id,
+                        )
+                    )
 
-            # If we have a limiter, garbage-collect all completed functions started
-            # before the period. If there's no limiter, there's no need--they were
-            # deleted on completion.
-            if queue.limiter is not None:
-                c.execute(
-                    sa.delete(SystemSchema.workflow_queue)
-                    .where(SystemSchema.workflow_queue.c.completed_at_epoch_ms != None)
-                    .where(SystemSchema.workflow_queue.c.queue_name == queue.name)
-                    .where(
-                        SystemSchema.workflow_queue.c.started_at_epoch_ms
-                        < start_time_ms - limiter_period_ms
+                    # Then give it a start time and assign the executor ID
+                    c.execute(
+                        SystemSchema.workflow_queue.update()
+                        .where(SystemSchema.workflow_queue.c.workflow_uuid == id)
+                        .values(
+                            started_at_epoch_ms=start_time_ms, executor_id=executor_id
+                        )
+                    )
+                    ret_ids.append(id)
+            except OperationalError as e:
+                # Abandon the queue items in case of serialization error
+                if isinstance(e.orig, errors.SerializationFailure):
+                    dbos_logger.warning("Serialization failure. Abandoning queue items")
+                    return []
+                else:
+                    # Re-raise other OperationalError types if unrelated
+                    raise
+            finally:
```

Comment:
Does this `finally` block actually work given that the transaction may have aborted in the catch?